### PR TITLE
Adding Indonesian Kartu Multi Trip support

### PR DESCRIPTION
**Pull request by sybond**
This PR is to support Kartu Multi Trip reading on Metrodroid. Anyway, I'm not sure if this is the correct way to fork changes from different branch.

### CODE REVIEW

**Comment by sybond** [comment]
Attached is sample of Kartu Multi Trip data.

[Metrodroid-Export_KMTSample.zip](https://github.com/micolous/metrodroid/files/2266259/Metrodroid-Export_KMTSample.zip)

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    private KMTUtil() {
+    }
+
+    static Calendar extractDate(byte[] data) {
+        int fulloffset = Util.toInt(data[0], data[1], data[2], data[3]);
+        if (fulloffset == 0) {
+            return null;
+        }
+        Calendar c = new GregorianCalendar(KMTTransitData.TIME_ZONE);
+        c.setTime(new Date((long) (fulloffset + 946684758) * 1000));
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java]: Please explicitly define epochs as a constant and as a date, as it makes it easier to read.  eg:

https://github.com/micolous/metrodroid/blob/9b8550615f8bf517fd5f7bd63ff188d7bf08ac26/src/main/java/au/id/micolous/metrodroid/transit/troika/TroikaTransitData.java#L58-L65

I'm not sure that the values you have here are exactly correct:

```
$ TZ=Asia/Jakarta date --date='@946684758'
Sat Jan  1 06:59:18 WIB 2000
$ TZ=UTC date --date='@946684758'
Fri Dec 31 23:59:18 UTC 1999
```

I think you're trying to set the epoch as 2000-01-01 00:00 UTC?

`ClipperTransitData` has a more complex example where the epoch is in UTC, but the rest of the card has dates in local time.

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+
+final class KMTUtil {
+    private KMTUtil() {
+    }
+
+    static Calendar extractDate(byte[] data) {
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java]: As there's only one method in `KMTUtil`, and it's only used with `KMTTrip`, maybe `extractDate` should just be a static method of that class?

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
+/*
+ * KMTTrip.java
+ *
+ * This file is part of FareBot.
+ * Learn more at: https://codebutler.github.io/farebot/
+ *
+ * Copyright (C) 2014-2016 Eric Butler <eric@codebutler.com>
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: You probably want your name in the copyright list ;)

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    };
+    private final int mProcessType;
+    private final int mSequenceNumber;
+    private final Calendar mTimestamp;
+    private final int mTransactionAmount;
+    private final int mEndGateCode;
+
+    public KMTTrip(FelicaBlock block) {
+        byte[] data = block.getData();
+        mProcessType = data[12];
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: `data[12] % 0xFF` will get rid of nasty surprises.

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    public String getEndStationName() {
+        // Need to work on decoding gate code
+        // Collect the data !!
+        return String.format("[%02X]",mEndGateCode);
+    }
+
+    public KMTTrip(Parcel parcel) {
+        mProcessType = parcel.readInt();
+        mSequenceNumber = parcel.readInt();
+        long t = parcel.readLong();
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: L73-79 can be replaced with:

```java
mTimestamp = Utils.longToCalendar(parcel.readLong(), KMTTransitData.TIME_ZONE);
```

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            mTimestamp = null;
+        }
+        mTransactionAmount = parcel.readInt();
+        mEndGateCode = parcel.readInt();
+    }
+
+    public void writeToParcel(Parcel parcel, int flags) {
+        parcel.writeInt(mProcessType);
+        parcel.writeInt(mSequenceNumber);
+        parcel.writeLong(mTimestamp == null ? 0 : mTimestamp.getTimeInMillis());
```
Comment (micolous) [inline src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: `parcel.writeLong(Utils.calendarToLong(mTimestamp));`

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java (hunk 4 above)
Comment (phcoder) [reply src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: & 0xff, not % 0xff

File: src/main/res/values/strings.xml [hunk 7]
```diff
@@ -674,6 +675,10 @@
     <string name="felica_process_payment_thirdparty">Payment (3rd Party)</string>
     <string name="felica_process_admission_thirdparty">Admission Payment (3rd Party)</string>
 
+    <!-- Kartu Multi Trip -->
+    <string name="kmt_def_route">JABODETABEK</string>
+    <string name="kmt_extra_note">Only for new FeliCa cards. Station data currently displayed as code.</string>
```
[the remark this answers is not in the document]
Comment (sybond) [reply src/main/res/values/strings.xml]: Default due to "unknown data meaning", so it better to display the generic route name "JABODETABEK" (area shortname that covered by Jakarta's Commuter Line services).

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java (hunk 1 above)
Comment (sybond) [reply src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java]: Thanks for pointing this out. Actually I'm not aware that epoch value are ` 2000-01-01 00:00 UTC`. I just manually calc real time during gate-out and see the diff in the data. I will check whether this epoch value is correct or not with the data.

Comment (sybond) [reply src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTUtil.java]: After playing around with some data, it seems spoch is not `2000-01-01 00:00 UTC`. Anyway following code will work, and easier to read :D

    static {
         GregorianCalendar epoch = new GregorianCalendar(TIME_ZONE);
         epoch.set(2000, 0, 1, 7, 0, 0);
         KMT_EPOCH = epoch.getTimeInMillis();
     }

Here is sample data: `22 F1 A7 E6`, and should be decoded as `30 Jul 2018 17:41 WIB`.

File: src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        else
+            str = app.getString(R.string.felica_terminal_pos);
+        return str;
+    }
+
+    public boolean hasTime() {
+        return mTimestamp != null;
+    }
+
+    public String getRouteName() {
```
[the remark this answers is not in the document]
Comment (sybond) [reply src/main/java/au/id/micolous/metrodroid/transit/kmt/KMTTrip.java]: Default due to "unknown data meaning", so it better to display the generic route name "JABODETABEK" (area shortname that covered by Jakarta's Commuter Line services).

**Comment by phcoder** [comment]
I pushed modified version to https://github.com/phcoder/metrodroid/tree/kmt . I tested it with your sample scans

**Comment by sybond** [comment]
Too much conflicts. I will close this PR, and rewrite based on latest update.

**Pull request closed without merging**